x-pack/legacy/plugins/lens/public/types.ts

```diff
@@ -159,17 +175,25 @@ export interface VisualizationSuggestion<T = unknown> {
   previewIcon: string;
 }
 
+export interface FramePublicAPI {
+  datasourceLayers: Record<string, DatasourcePublicAPI>;
+  // Adds a new layer. This triggers a re-render
+  addNewLayer: () => string;
+}
+
 export interface Visualization<T = unknown, P = unknown> {
   // For initializing from saved object
-  initialize: (datasource: DatasourcePublicAPI, state?: P) => T;
+  initialize: (frame: FramePublicAPI, state?: P) => T;
 
   getPersistableState: (state: T) => P;
 
   renderConfigPanel: (domElement: Element, props: VisualizationProps<T>) => void;
 
-  toExpression: (state: T, datasource: DatasourcePublicAPI) => Ast | string | null;
+  toExpression: (state: T, frame: FramePublicAPI) => Ast | string | null;
 
   // The frame will call this function on all visualizations when the table changes, or when
   // rendering additional ways of using the data
   getSuggestions: (options: SuggestionRequest<T>) => Array<VisualizationSuggestion<T>>;
+
+  getLayerIds: (state: T) => string[];
```

Comment:
We are not using those, right? The datasources return the layers, not the visualization.